src/idl/src/annotation.c

```diff
@@ -510,6 +510,299 @@ annotate_must_understand(
+      case IDL_UINT8:
+      case IDL_OCTET:
+        range_error =
+          (is_unsigned && uval > UINT8_MAX) ||
+          (!is_unsigned && (sval > UINT8_MAX || sval < 0));
+        break;
+      case IDL_SHORT:
+      case IDL_INT16:
+        range_error =
+          (is_unsigned && uval > INT16_MAX) ||
+          (!is_unsigned && (sval > INT16_MAX || sval < INT16_MIN));
+        break;
+      case IDL_UINT16:
+      case IDL_USHORT:
+        range_error =
+          (is_unsigned && uval > UINT16_MAX) ||
+          (!is_unsigned && (sval > UINT16_MAX || sval < 0));
+        break;
+      case IDL_INT32:
+      case IDL_LONG:
+        range_error =
+          (is_unsigned && uval > INT32_MAX) ||
+          (!is_unsigned && (sval > INT32_MAX || sval < INT32_MIN));
+        break;
+      case IDL_UINT32:
+      case IDL_ULONG:
+        range_error =
+          (is_unsigned && uval > UINT32_MAX) ||
+          (!is_unsigned && (sval > UINT32_MAX || sval < 0));
+        break;
+      case IDL_LLONG:
+      case IDL_INT64:
+        range_error = is_unsigned && uval > INT64_MAX;
+        break;
+      case IDL_ULLONG:
+      case IDL_UINT64:
+        range_error = !is_unsigned && sval < 0;
+        break;
+      default:
+        //boundary checking done on setting int limits on floating point fields
+        break;
+    }
+
+    if (range_error) {
+      idl_error(pstate, idl_location(const_expr),
+        "@max/@min/@range limit range error");
+      return IDL_RETCODE_SEMANTIC_ERROR;
+    }
+  } else if (param_type & IDL_FLOATING_PT_TYPE) {
+    if (!(field_type & IDL_FLOATING_PT_TYPE)) {
+      idl_error(pstate, idl_location(const_expr),
+        "@max/@min/@range error: floating point limit on non-floating point field");
+      return IDL_RETCODE_SEMANTIC_ERROR;
+    }
+  } else {
+    idl_error(pstate, idl_location(param),
+      "@max/@min/@range with %s cannot be applied to '%s' element",
+      idl_construct(const_expr), idl_construct(node));
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  //check that min is not larger than max
+  const idl_literal_t *min_lit = NULL,
+                      *max_lit = NULL;
+  if (idl_is_member(node)) {
+    if (to_max) {
+      ((idl_member_t*)node)->max.annotation = parent;
+      ((idl_member_t*)node)->max.value = lit;
+    } else {
+      ((idl_member_t*)node)->min.annotation = parent;
+      ((idl_member_t*)node)->min.value = lit;
+    }
+    min_lit = ((idl_member_t*)node)->min.value;
+    max_lit = ((idl_member_t*)node)->max.value;
+  } else if (idl_is_case(node)) {
+    if (to_max) {
+      ((idl_case_t*)node)->max.annotation = parent;
+      ((idl_case_t*)node)->max.value = lit;
+    } else {
+      ((idl_case_t*)node)->min.annotation = parent;
+      ((idl_case_t*)node)->min.value = lit;
+    }
+    min_lit = ((idl_case_t*)node)->min.value;
+    max_lit = ((idl_case_t*)node)->max.value;
+  }
+
+  if (min_lit && max_lit && idl_arithmetic_to_double(min_lit) > idl_arithmetic_to_double(max_lit)) {
```

Comment:
All the checking is done with such great care (much appreciated), but then unfortunately there is this ... double-precision floating-point has a mantissa of 53 bits and so you get loss of precision when converting 64-bit integers of magnitude > 2^53 (or thereabouts ...) to floating-point.

Try it with 2^64-1 and 2^64-2, for example: they convert to the same floating-point number.

```C
#include <stdint.h>
#include <stdio.h>
#include <inttypes.h>
int main(int argc,char **argv){
  uint64_t a, b;
  sscanf(argv[1],"%"SCNu64,&a);
  sscanf(argv[2],"%"SCNu64,&b);
  printf("%a > %a = %d\n",(double)a,(double)b,(double)a>(double)b);
  return 0;
}
```

Reply:
This occurs because the two parameters are parsed separately, whether in the same annotation like @range, or with @min and @max, and one can be read as an int, and the other as a float.
I will put in a small check that will give a warning if you compare integers and floats in a manner that loses resolution.